### PR TITLE
Init Commit

### PR DESCRIPTION
**Pull request by d1onaea**

### CODE REVIEW

File: PinkPong/Assets/Scripts/Ball.cs [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 public class Ball : MonoBehaviour
 {
+    [SerializeField]
+    private GameManager gameManager;
+
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs]: Incorrect dependency.
`GameManager` can know something about `Ball` but not vice versa.
Make similar field in `GameManager` and remove this field.

Comment (Kiuh) [reply PinkPong/Assets/Scripts/Ball.cs]: And in other scripts too

File: PinkPong/Assets/Scripts/GameManager.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    void Start()
+    {
+        
+    }
+
+    // Update is called once per frame
+    void Update()
+    {
+        
+    }
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/GameManager.cs]: Don't leave empty methods, it's dead code

File: PinkPong/Assets/Scripts/Paddle.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+    [SerializeField] private Rigidbody2D rb2d;
+
+    [SerializeField] private string inputButton;
+
+    [SerializeField] private float speed;
+
+    [SerializeField] private Transform target;
+
+    [SerializeField] private GameManager gameManager;
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Paddle.cs:14]: Apply formatter, like it was done in `Ball` script.

Comment (Kiuh) [reply PinkPong/Assets/Scripts/Paddle.cs:14]: Better apply formatter to all scripts

File: PinkPong/Assets/Scripts/Paddle.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        if (Input.GetButton(inputButton))
+        {
+            Move(Input.GetAxis(inputButton));
+        }
+    }
+
+    private void Move(float movement)
+    {
+        transform.RotateAround(target.position, new Vector3(0, 0, 1) * movement, Time.deltaTime * speed);
+    }
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Paddle.cs]: Try separate this logic in two scripts, one of them will manage only input of paddle and other one only paddle movement.

Comment (Kiuh) [reply PinkPong/Assets/Scripts/Paddle.cs]: Better introduce script that will manage all user input.

**Review by Kiuh — CHANGES_REQUESTED** [review]

File: PinkPong/Assets/Scripts/PaddleMovement.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    public void Move(float movement)
+    {
+        transform.RotateAround(target.position, new Vector3(0, 0, 1) * movement, Time.deltaTime * speed);
+    }
+
+    private void Dash(float movement)
+    {
+        transform.RotateAround(target.position, new Vector3(0, 0, 1) * movement, Time.deltaTime * dashSpeed);
+    }
```
Comment (mertwole) [inline PinkPong/Assets/Scripts/PaddleMovement.cs:41]:
```suggestion
    private void Dash(float movement)
    {
        transform.RotateAround(target.position, Vector3.forward * movement, Time.deltaTime * dashSpeed);
    }
```

**Review by mertwole — CHANGES_REQUESTED** [review]

File: PinkPong/Assets/Scripts/PaddleMovement.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    private Vector3 startPosition;
+
+    private void Start()
+    {
+        startPosition = transform.position;
+    }
+
+    private void OnEnable()
+    {
+        input.IsDoublePressed += Dash;
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/PaddleMovement.cs:20]: Don't use events here, just use `Dash` the same way as `Move` - call it from `PaddleInput`

File: PinkPong/Assets/Scripts/PaddleMovement.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
+using UnityEngine;
+
+public class PaddleMovement : MonoBehaviour
+{
+    [SerializeField] private Rigidbody2D rb2d;
+    [SerializeField] private PaddleInput input;
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/PaddleMovement.cs:6]:
```suggestion
```

File: PinkPong/Assets/Scripts/PaddleMovement.cs [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        transform.position = startPosition;
+    }
+
+    public void Move(float movement)
+    {
+        transform.RotateAround(target.position, new Vector3(0, 0, 1) * movement, Time.deltaTime * speed);
+    }
+
+    private void Dash(float movement)
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/PaddleMovement.cs:38]: Assure that dash mechanic is working properly - it should teleport paddle instead of speeding it up

Comment (Kiuh) [reply PinkPong/Assets/Scripts/PaddleMovement.cs:38]: And, of course, it shouldn't depend on `Time.deltaTime`

File: PinkPong/Assets/Scripts/Ball.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+//    private void OnCollisionEnter2D(Collision2D collision)
+//    {
+//        Paddle paddle = collision.collider.GetComponent<Paddle>();
+//        if (paddle)
+//        {
+//            rb2d.velocity *= speedMultiplier;
+//        }
+//    }
+//}
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs:72]:
```suggestion
```
Don't leave commented out code in git

File: PinkPong/Assets/Scripts/Ball.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    private void ResetBallPosition()
+    {
+        float posY = Random.Range(-maxStartY, maxStartY);
+        Vector2 position = new(startX, posY);
+        transform.position = position;
+    }
+}
+
+//� �� ��� ��� ������� ��������� ������
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs:39]: :grin:

File: PinkPong/Assets/Scripts/Ball.cs
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs:1]: Please re-check that formatter is working

File: PinkPong/Assets/Scripts/Ball.cs [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    private readonly float startX = 0;
+    private readonly float maxStartY = 4f;
+
+    private void Start()
+    {
+        InitialPush();
+        GameManager.Instance.OnReset += ResetBall;
+    }
+
+    private void InitialPush()
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs:18]: Change the logics to actually generating normalized direction and the multiplying it to `moveSpeed`

Comment (Kiuh) [reply PinkPong/Assets/Scripts/Ball.cs:18]: Use trigonometry for generating normalized direction

File: PinkPong/Assets/Scripts/Ball.cs [hunk 12]
```diff
@@ -0,0 +1,72 @@
+using UnityEngine;
+
+public class Ball : MonoBehaviour
+{
+    [SerializeField] private Rigidbody2D rb2d;
+    [SerializeField] private float maxInitAngle = 0.67f;
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs:6]: It woulde more convinient if this value will be specified in degrees

File: PinkPong/Assets/Scripts/Ball.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        rb2d.velocity = direction * moveSpeed;
+    }
+
+    private void ResetBall()
+    {
+        ResetBallPosition();
+        InitialPush();
+    }
+
+    private void ResetBallPosition()
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs:31]: The logics here is different from one discussed

File: PinkPong/Assets/Scripts/PaddleInput.cs [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+    [SerializeField] private PaddleMovement movement;
+    [SerializeField] private float delayBetweenPresses;
+    private bool pressedFirstTime = false;
+    private float lastPressedTime;
+
+    public event Action<float> IsDoublePressed;
+
+    private void Update()
+    {
+        if (Input.GetButtonUp(inputButton))
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/PaddleInput.cs:16]:
```suggestion
        if (Input.GetButtonDown(inputButton))
```

File: PinkPong/Assets/Scripts/PaddleInput.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        lastPressedTime = Time.time;
+        pressedFirstTime = !pressedFirstTime;
+
+        if (pressedFirstTime && Time.time - lastPressedTime > delayBetweenPresses)
+        {
+            pressedFirstTime = !pressedFirstTime;
+        }
+    }
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/PaddleInput.cs:45]:
```suggestion
    private bool DoublePressCheck()
    {
         var doublePressed = Time.time - lastPressedTime > delayBetweenPresses;
         lastPressedTime = Time.time;
         return doublePressed;
    }
```

File: PinkPong/Assets/Scripts/PaddleMovement.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        input.IsDoublePressed -= Dash;
+    }
+
+    public void ResetPaddle()
+    {
+        transform.position = startPosition;
+    }
+
+    public void Move(float movement)
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/PaddleMovement.cs:33]:
```suggestion
    public void Move(RotationDirection direction)
```

where `RotationDirection` is 
```c#
enum RotationDirection {
    Clockwise,
    CounterClockWise
}
```

File: PinkPong/Assets/Scripts/PaddleMovement.cs [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        transform.RotateAround(target.position, new Vector3(0, 0, 1) * movement, Time.deltaTime * speed);
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/PaddleMovement.cs:35]: This approach is counter-intuitive, better multiply the third argument to direction. Select multiplier in accordance to the `RotationDirection` enum value by `switch`

Comment (Kiuh) [inline PinkPong/Assets/Scripts/PaddleMovement.cs:35]:
```suggestion
        transform.RotateAround(target.position, Vector3.forward * movement, Time.deltaTime * speed);
```

File: PinkPong/Assets/Scripts/Ball.cs [hunk 18]
```diff
@@ -0,0 +1,72 @@
+using UnityEngine;
+
+public class Ball : MonoBehaviour
+{
+    [SerializeField] private Rigidbody2D rb2d;
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs:5]:
```suggestion
    [SerializeField] private Rigidbody2D rigidbody;
```

File: PinkPong/Assets/Scripts/Ball.cs [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+    [SerializeField] private float maxInitAngle = 0.67f;
+    [Range(0, 10)][SerializeField] private float moveSpeed = 1f;
+    [SerializeField] private float speedMultiplier = 1.1f;
+    private readonly float startX = 0;
+    private readonly float maxStartY = 4f;
```
Comment (Kiuh) [inline PinkPong/Assets/Scripts/Ball.cs:10]: `[SerializeField] private ...` approach here

Comment (d1onaea) [reply PinkPong/Assets/Scripts/Ball.cs:10]: why? it's not like you need to change the value in the inspector

Comment (Kiuh) [reply PinkPong/Assets/Scripts/Ball.cs:10]: because if something changes in environment, you will go to the code to change it.

**Review by Kiuh — APPROVED** [review]

**Pull request merged**